kv/kvserver: TestStoreRangeMergeRaftSnapshot failed

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://tanzanite.cluster.engflow.com/invocations/default/1da0f0c1-be1e-46e5-9c9c-6041a188223a?testReportRun=1&testReportShard=18&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [f9918d8f81a1829df63ac734fd6d21c60141e338](https://github.com/cockroachdb/cockroach/commits/f9918d8f81a1829df63ac734fd6d21c60141e338):

Fatal error:

```
panic: KeyComparison of ("\x03\xfa\x00\x00\x00\x18\x03\xd3\"\x8c\xf0\x8f\xbe\t", "\x03\xfa\x00\x18\x03\xd3\"\x84\xfc3\x91\t") = (prefix={5,common=3} cmp=1), ComparePrev gave (prefix={5,common=2} cmp=1)

```
Stack: 

```
goroutine 11803 [running]:
github.com/cockroachdb/pebble/sstable/colblk.AssertKeyCompare(0x134851a0, {0xc00ecc7c80, 0xe, 0x80}, {0xc00915dc00, 0xc, 0x40}, {0x5, 0x2, 0x1})
	external/com_github_cockroachdb_pebble/sstable/colblk/data_block.go:113 +0x40f
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).evaluatePoint(0xc001360608, {{0xc00ecc7c80?, 0x3?, 0x80?}, 0xc007b4e000?}, 0x4a)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:477 +0x165
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).AddWithForceObsolete(0xc001360608, {{0xc00ecc7c80?, 0xe?, 0x80?}, 0x30989?}, {0xc007b4e000, 0x4a, 0x60000}, 0x0)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:347 +0x198
github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable(0xc006786e88, {0xecb0200, 0xc001360608})
	external/com_github_cockroachdb_pebble/internal/compact/run.go:203 +0xa7c
github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable(0xc006786e88, {0x9, 0x2, {0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0}, ...}}, ...)
	external/com_github_cockroachdb_pebble/internal/compact/run.go:143 +0x26f
github.com/cockroachdb/pebble.(*DB).compactAndWrite(0xc00e0d0008, 0x8, 0xc0094af208, {0x0, 0x0, 0x0}, 0x7)
	external/com_github_cockroachdb_pebble/compaction.go:2913 +0xdc6
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc00e0d0008, 0x8, 0xc0094af208)
	external/com_github_cockroachdb_pebble/compaction.go:2808 +0xc45
github.com/cockroachdb/pebble.(*DB).flush1(0xc00e0d0008)
	external/com_github_cockroachdb_pebble/compaction.go:1448 +0x88d
github.com/cockroachdb/pebble.(*DB).flush.func1({0xec3ad38?, 0xc00df69500?})
	external/com_github_cockroachdb_pebble/compaction.go:1183 +0x145
runtime/pprof.Do({0xec3a448, 0x14b4a140}, {{0xc000ec1140?, 0x1f3b6c0?, 0xc00ac88e60?}}, 0xc009a74fa0)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
github.com/cockroachdb/pebble.(*DB).flush(0xc00e0d0008)
	external/com_github_cockroachdb_pebble/compaction.go:1176 +0x92
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush in goroutine 7268
	external/com_github_cockroachdb_pebble/compaction.go:1092 +0x187
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestStoreRangeMergeRaftSnapshot3234586533
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=18</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-43902

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://tanzanite.cluster.engflow.com/invocations/default/09b86016-ea5b-4088-b159-d74215970441?testReportRun=3&testReportShard=18&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [17535c13cfed95db70cd8dfb1ba6a700686f57b1](https://github.com/cockroachdb/cockroach/commits/17535c13cfed95db70cd8dfb1ba6a700686f57b1):

Fatal error:

```
panic: KeyComparison of ("\x03\xfa\x00\x00\x00\x18\x04\">\x02\xb1v\xe7\t", "\x03\xfa\x00\x18\x04\"=\xf4\xbbk\xa5\t") = (prefix={5,common=3} cmp=1), ComparePrev gave (prefix={5,common=2} cmp=1)

```
Stack: 

```
goroutine 11488 [running]:
github.com/cockroachdb/pebble/sstable/colblk.AssertKeyCompare(0x1348c200, {0xc001cf9680, 0xe, 0x80}, {0xc002814b80, 0xc, 0x40}, {0x5, 0x2, 0x1})
	external/com_github_cockroachdb_pebble/sstable/colblk/data_block.go:113 +0x40f
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).evaluatePoint(0xc00370ac08, {{0xc001cf9680?, 0x3?, 0x80?}, 0xc00ded4000?}, 0x43)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:477 +0x165
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).AddWithForceObsolete(0xc00370ac08, {{0xc001cf9680?, 0xe?, 0x80?}, 0x2f87a?}, {0xc00ded4000, 0x43, 0x60000}, 0x0)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:347 +0x198
github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable(0xc00dd10e88, {0xecb5520, 0xc00370ac08})
	external/com_github_cockroachdb_pebble/internal/compact/run.go:203 +0xa7c
github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable(0xc00dd10e88, {0x9, 0x2, {0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0}, ...}}, ...)
	external/com_github_cockroachdb_pebble/internal/compact/run.go:143 +0x26f
github.com/cockroachdb/pebble.(*DB).compactAndWrite(0xc00dca5208, 0x8, 0xc0090b4308, {0x0, 0x0, 0x0}, 0x7)
	external/com_github_cockroachdb_pebble/compaction.go:2913 +0xdc6
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc00dca5208, 0x8, 0xc0090b4308)
	external/com_github_cockroachdb_pebble/compaction.go:2808 +0xc45
github.com/cockroachdb/pebble.(*DB).flush1(0xc00dca5208)
	external/com_github_cockroachdb_pebble/compaction.go:1448 +0x88d
github.com/cockroachdb/pebble.(*DB).flush.func1({0xec3ff58?, 0xc00a6ffad0?})
	external/com_github_cockroachdb_pebble/compaction.go:1183 +0x145
runtime/pprof.Do({0xec3f668, 0x14b51240}, {{0xc001547640?, 0x1f3b8a0?, 0xc00e034fd0?}}, 0xc00e034fa0)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
github.com/cockroachdb/pebble.(*DB).flush(0xc00dca5208)
	external/com_github_cockroachdb_pebble/compaction.go:1176 +0x92
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush in goroutine 6829
	external/com_github_cockroachdb_pebble/compaction.go:1092 +0x187
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestStoreRangeMergeRaftSnapshot3781572371
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=18</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://tanzanite.cluster.engflow.com/invocations/default/099d2f6d-4458-4022-bc54-7332ff70444c?testReportRun=1&testReportShard=18&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [6bb6dc96ebf0ee2f23c5c568fa0d421019dc0946](https://github.com/cockroachdb/cockroach/commits/6bb6dc96ebf0ee2f23c5c568fa0d421019dc0946):

Fatal error:

```
panic: KeyComparison of ("\x03\xfa\x00\x00\x00\x18\x04t\xa6\xa9D\x88!\t", "\x03\xfa\x00\x18\x04t\xa6\xa0t\xe9\x17\t") = (prefix={5,common=3} cmp=1), ComparePrev gave (prefix={5,common=2} cmp=1)

```
Stack: 

```
goroutine 11695 [running]:
github.com/cockroachdb/pebble/sstable/colblk.AssertKeyCompare(0x13490200, {0xc002611180, 0xe, 0x80}, {0xc002611380, 0xc, 0x80}, {0x5, 0x2, 0x1})
	external/com_github_cockroachdb_pebble/sstable/colblk/data_block.go:113 +0x40f
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).evaluatePoint(0xc00b051908, {{0xc002611180?, 0x3?, 0x80?}, 0xc009990000?}, 0x43)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:477 +0x165
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).AddWithForceObsolete(0xc00b051908, {{0xc002611180?, 0xe?, 0x80?}, 0x2f728?}, {0xc009990000, 0x43, 0x60000}, 0x0)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:347 +0x198
github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable(0xc007726e88, {0xecb8080, 0xc00b051908})
	external/com_github_cockroachdb_pebble/internal/compact/run.go:203 +0xa7c
github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable(0xc007726e88, {0x9, 0x2, {0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0}, ...}}, ...)
	external/com_github_cockroachdb_pebble/internal/compact/run.go:143 +0x26f
github.com/cockroachdb/pebble.(*DB).compactAndWrite(0xc003ad2008, 0x8, 0xc0054db508, {0x0, 0x0, 0x0}, 0x7)
	external/com_github_cockroachdb_pebble/compaction.go:2913 +0xdc6
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc003ad2008, 0x8, 0xc0054db508)
	external/com_github_cockroachdb_pebble/compaction.go:2808 +0xc45
github.com/cockroachdb/pebble.(*DB).flush1(0xc003ad2008)
	external/com_github_cockroachdb_pebble/compaction.go:1448 +0x88d
github.com/cockroachdb/pebble.(*DB).flush.func1({0xec42ab8?, 0xc00d86d9b0?})
	external/com_github_cockroachdb_pebble/compaction.go:1183 +0x145
runtime/pprof.Do({0xec421c8, 0x14b55280}, {{0xc000e6d280?, 0x1f3b840?, 0xc008bc0b40?}}, 0xc0089c8fa0)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
github.com/cockroachdb/pebble.(*DB).flush(0xc003ad2008)
	external/com_github_cockroachdb_pebble/compaction.go:1176 +0x92
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush in goroutine 7090
	external/com_github_cockroachdb_pebble/compaction.go:1092 +0x187
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestStoreRangeMergeRaftSnapshot2210132588
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=18</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://tanzanite.cluster.engflow.com/invocations/default/b1ce1a73-3cf0-4f83-96db-b4e555ad6751?testReportRun=1&testReportShard=18&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [546c489ae32a83b8d312977156cc8f2be93b6a49](https://github.com/cockroachdb/cockroach/commits/546c489ae32a83b8d312977156cc8f2be93b6a49):

Fatal error:

```
panic: KeyComparison of ("\x03\xfa\x00\x00\x00\x18\x04\xc2\x7f\a\xe4\x14!\t", "\x03\xfa\x00\x18\x04\xc2\x7f\x01\x94\x1c\xb6\t") = (prefix={5,common=3} cmp=1), ComparePrev gave (prefix={5,common=2} cmp=1)

```
Stack: 

```
goroutine 10994 [running]:
github.com/cockroachdb/pebble/sstable/colblk.AssertKeyCompare(0x1348f240, {0xc00426c080, 0xe, 0x80}, {0xc009dfdb80, 0xc, 0x40}, {0x5, 0x2, 0x1})
	external/com_github_cockroachdb_pebble/sstable/colblk/data_block.go:113 +0x40f
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).evaluatePoint(0xc00769b308, {{0xc00426c080?, 0x3?, 0x80?}, 0xc00c57a000?}, 0x43)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:477 +0x165
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).AddWithForceObsolete(0xc00769b308, {{0xc00426c080?, 0xe?, 0x80?}, 0x2f1e8?}, {0xc00c57a000, 0x43, 0x60000}, 0x0)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:347 +0x198
github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable(0xc00c222e88, {0xecb6a60, 0xc00769b308})
	external/com_github_cockroachdb_pebble/internal/compact/run.go:203 +0xa7c
github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable(0xc00c222e88, {0x9, 0x2, {0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0}, ...}}, ...)
	external/com_github_cockroachdb_pebble/internal/compact/run.go:143 +0x26f
github.com/cockroachdb/pebble.(*DB).compactAndWrite(0xc009de4d88, 0x8, 0xc0036bb808, {0x0, 0x0, 0x0}, 0x7)
	external/com_github_cockroachdb_pebble/compaction.go:2913 +0xdc6
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc009de4d88, 0x8, 0xc0036bb808)
	external/com_github_cockroachdb_pebble/compaction.go:2808 +0xc45
github.com/cockroachdb/pebble.(*DB).flush1(0xc009de4d88)
	external/com_github_cockroachdb_pebble/compaction.go:1448 +0x88d
github.com/cockroachdb/pebble.(*DB).flush.func1({0xec41498?, 0xc0090000c0?})
	external/com_github_cockroachdb_pebble/compaction.go:1183 +0x145
runtime/pprof.Do({0xec40ba8, 0x14b542c0}, {{0xc0000b71c0?, 0x1f3bd40?, 0xc0051adfd0?}}, 0xc0051adfa0)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
github.com/cockroachdb/pebble.(*DB).flush(0xc009de4d88)
	external/com_github_cockroachdb_pebble/compaction.go:1176 +0x92
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush in goroutine 6909
	external/com_github_cockroachdb_pebble/compaction.go:1092 +0x187
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestStoreRangeMergeRaftSnapshot647396015
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=18</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://tanzanite.cluster.engflow.com/invocations/default/a664d1bf-f30d-4b79-afa3-4cbd3d2a1187?testReportRun=1&testReportShard=18&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [e8f6410ecd99591fd6103dd8461ea3fb0361f8d6](https://github.com/cockroachdb/cockroach/commits/e8f6410ecd99591fd6103dd8461ea3fb0361f8d6):

Fatal error:

```
panic: KeyComparison of ("\x03\xfa\x00\x00\x00\x18\x05\x12ͷ\x1f3\xf2\t", "\x03\xfa\x00\x18\x05\x12ͪN\x9f\xde\x00\x00\x00\x01\r") = (prefix={5,common=3} cmp=1), ComparePrev gave (prefix={5,common=2} cmp=1)

```
Stack: 

```
goroutine 11540 [running]:
github.com/cockroachdb/pebble/sstable/colblk.AssertKeyCompare(0x13478240, {0xc00452bf00, 0xe, 0x80}, {0xc002880340, 0x10, 0x40}, {0x5, 0x2, 0x1})
	external/com_github_cockroachdb_pebble/sstable/colblk/data_block.go:113 +0x40f
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).evaluatePoint(0xc004743908, {{0xc00452bf00?, 0x3?, 0x80?}, 0xc00ad4a000?}, 0x43)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:477 +0x165
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).AddWithForceObsolete(0xc004743908, {{0xc00452bf00?, 0xe?, 0x80?}, 0x2f55f?}, {0xc00ad4a000, 0x43, 0x62000}, 0x0)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:347 +0x198
github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable(0xc00a926e88, {0xeca4680, 0xc004743908})
	external/com_github_cockroachdb_pebble/internal/compact/run.go:203 +0xa7c
github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable(0xc00a926e88, {0x9, 0x2, {0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0}, ...}}, ...)
	external/com_github_cockroachdb_pebble/internal/compact/run.go:143 +0x26f
github.com/cockroachdb/pebble.(*DB).compactAndWrite(0xc0053ac008, 0x8, 0xc002de8c08, {0x0, 0x0, 0x0}, 0x7)
	external/com_github_cockroachdb_pebble/compaction.go:2907 +0xdc6
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc0053ac008, 0x8, 0xc002de8c08)
	external/com_github_cockroachdb_pebble/compaction.go:2802 +0xc45
github.com/cockroachdb/pebble.(*DB).flush1(0xc0053ac008)
	external/com_github_cockroachdb_pebble/compaction.go:1442 +0x88d
github.com/cockroachdb/pebble.(*DB).flush.func1({0xec2f1b8?, 0xc008fa1740?})
	external/com_github_cockroachdb_pebble/compaction.go:1180 +0x145
runtime/pprof.Do({0xec2e8c8, 0x14b3e6c0}, {{0xc0000b6dc0?, 0x1f3b260?, 0xc00985b7c0?}}, 0xc00ebb9fa0)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
github.com/cockroachdb/pebble.(*DB).flush(0xc0053ac008)
	external/com_github_cockroachdb_pebble/compaction.go:1173 +0x92
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush in goroutine 6988
	external/com_github_cockroachdb_pebble/compaction.go:1089 +0x187
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestStoreRangeMergeRaftSnapshot1583420874
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=18</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://tanzanite.cluster.engflow.com/invocations/default/53efa5c8-5982-4049-8d31-516dc6d35af4?testReportRun=2&testReportShard=18&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [f5fde858b71c965da8068e211e46d2a3d297e876](https://github.com/cockroachdb/cockroach/commits/f5fde858b71c965da8068e211e46d2a3d297e876):

Fatal error:

```
panic: KeyComparison of ("\x03\xfa\x00\x00\x00\x18\x05^\x96\xc2\xe9\xb7\xe0\t", "\x03\xfa\x00\x18\x05^\x96\xbb54>\t") = (prefix={5,common=3} cmp=1), ComparePrev gave (prefix={5,common=2} cmp=1)

```
Stack: 

```
goroutine 11408 [running]:
github.com/cockroachdb/pebble/sstable/colblk.AssertKeyCompare(0x135a1240, {0xc00edff800, 0xe, 0x80}, {0xc00a1f3740, 0xc, 0x40}, {0x5, 0x2, 0x1})
	external/com_github_cockroachdb_pebble/sstable/colblk/data_block.go:113 +0x40f
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).evaluatePoint(0xc004bed308, {{0xc00edff800?, 0x3?, 0x80?}, 0xc00bc78000?}, 0x43)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:477 +0x165
github.com/cockroachdb/pebble/sstable.(*RawColumnWriter).AddWithForceObsolete(0xc004bed308, {{0xc00edff800?, 0xe?, 0x80?}, 0x2f22e?}, {0xc00bc78000, 0x43, 0x62000}, 0x0)
	external/com_github_cockroachdb_pebble/sstable/colblk_writer.go:347 +0x198
github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable(0xc00ba1ee88, {0xed9e360, 0xc004bed308})
	external/com_github_cockroachdb_pebble/internal/compact/run.go:203 +0xa7c
github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable(0xc00ba1ee88, {0x9, 0x2, {0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0}, ...}}, ...)
	external/com_github_cockroachdb_pebble/internal/compact/run.go:143 +0x26f
github.com/cockroachdb/pebble.(*DB).compactAndWrite(0xc0089b0008, 0x8, 0xc009a43808, {0x0, 0x0, 0x0}, 0x7)
	external/com_github_cockroachdb_pebble/compaction.go:2907 +0xdc6
github.com/cockroachdb/pebble.(*DB).runCompaction(0xc0089b0008, 0x8, 0xc009a43808)
	external/com_github_cockroachdb_pebble/compaction.go:2802 +0xc45
github.com/cockroachdb/pebble.(*DB).flush1(0xc0089b0008)
	external/com_github_cockroachdb_pebble/compaction.go:1442 +0x88d
github.com/cockroachdb/pebble.(*DB).flush.func1({0xed28dd8?, 0xc00e4c5dd0?})
	external/com_github_cockroachdb_pebble/compaction.go:1180 +0x145
runtime/pprof.Do({0xed284e8, 0x14c67740}, {{0xc000ebe600?, 0x1f3b600?, 0xc0047c4fd0?}}, 0xc0047c4fa0)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
github.com/cockroachdb/pebble.(*DB).flush(0xc0089b0008)
	external/com_github_cockroachdb_pebble/compaction.go:1173 +0x92
created by github.com/cockroachdb/pebble.(*DB).maybeScheduleFlush in goroutine 6960
	external/com_github_cockroachdb_pebble/compaction.go:1089 +0x187
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestStoreRangeMergeRaftSnapshot810168479
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=18</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/storage/pebble_key_schema.go
+++ b/pkg/storage/pebble_key_schema.go
@@ -97,6 +97,7 @@ type cockroachKeyWriter struct {
 	logicalTimes    colblk.UintBuilder
 	untypedVersions colblk.RawBytesBuilder
 	suffixTypes     suffixTypes
+	prevRoachKeyLen int32
 	prevSuffix      []byte
 }
 
@@ -118,35 +119,57 @@ func (kw *cockroachKeyWriter) Reset() {
 	kw.logicalTimes.Reset()
 	kw.untypedVersions.Reset()
 	kw.suffixTypes = 0
+	kw.prevRoachKeyLen = 0
 }
 
 func (kw *cockroachKeyWriter) ComparePrev(key []byte) colblk.KeyComparison {
-	var cmpv colblk.KeyComparison
-	cmpv.PrefixLen = int32(EngineKeySplit(key)) // TODO(jackson): Inline
+	prefixLen := EngineKeySplit(key)
 	if kw.roachKeys.Rows() == 0 {
-		cmpv.UserKeyComparison = 1
-		return cmpv
+		return colblk.KeyComparison{
+			PrefixLen:         int32(prefixLen),
+			CommonPrefixLen:   0,
+			UserKeyComparison: +1,
+		}
 	}
-	lp := kw.roachKeys.UnsafeGet(kw.roachKeys.Rows() - 1)
-	cmpv.CommonPrefixLen = int32(crbytes.CommonPrefix(lp, key[:cmpv.PrefixLen-1]))
-	if cmpv.CommonPrefixLen == cmpv.PrefixLen-1 {
-		// Adjust CommonPrefixLen to include the sentinel byte.
-		cmpv.CommonPrefixLen = cmpv.PrefixLen
-		cmpv.UserKeyComparison = int32(EnginePointSuffixCompare(key[cmpv.PrefixLen:], kw.prevSuffix))
-		return cmpv
+	lastRoachKey := kw.roachKeys.UnsafeGet(kw.roachKeys.Rows() - 1)
+	commonPrefixLen := crbytes.CommonPrefix(lastRoachKey, key[:prefixLen-1])
+	if len(lastRoachKey) == commonPrefixLen {
+		if buildutil.CrdbTestBuild && len(lastRoachKey) > prefixLen-1 {
+			panic(errors.AssertionFailedf("out-of-order keys: previous roach key %q > roach key of key %q",
+				lastRoachKey, key))
+		}
+		// All the bytes of the previous roach key form a byte-wise prefix of
+		// [key]'s prefix. The last byte of the previous prefix is the 0x00
+		// sentinel byte, which is not stored within roachKeys. It's possible
+		// that [key] also has a 0x00 byte in the same position (either also
+		// serving as a sentinel byte, in which case the prefixes are equal, or
+		// not in which case [key] is greater). In both cases, we need to
+		// increment CommonPrefixLen.
+		if key[commonPrefixLen] == 0x00 {
+			commonPrefixLen++
+			if commonPrefixLen == prefixLen {
+				// The prefixes are equal; compare the suffixes.
+				return colblk.KeyComparison{
+					PrefixLen:         int32(prefixLen),
+					CommonPrefixLen:   int32(commonPrefixLen),
+					UserKeyComparison: int32(EnginePointSuffixCompare(key[prefixLen:], kw.prevSuffix)),
+				}
+			}
+		}
+		// prefixLen > commonPrefixLen; key is greater.
+		return colblk.KeyComparison{
+			PrefixLen:         int32(prefixLen),
+			CommonPrefixLen:   int32(commonPrefixLen),
+			UserKeyComparison: +1,
+		}
 	}
-	// The keys have different MVCC prefixes. We haven't determined which is
-	// greater, but we know the index at which they diverge. The base.Comparer
-	// contract dictates that prefixes must be lexicographically ordered.
-	if len(lp) == int(cmpv.CommonPrefixLen) {
-		// cmpv.PrefixLen > cmpv.PrefixLenShared; key is greater.
-		cmpv.UserKeyComparison = +1
-	} else {
-		// Both keys have at least 1 additional byte at which they diverge.
-		// Compare the diverging byte.
-		cmpv.UserKeyComparison = int32(cmp.Compare(key[cmpv.CommonPrefixLen], lp[cmpv.CommonPrefixLen]))
+	// Both keys have at least 1 additional byte at which they diverge.
+	// Compare the diverging byte.
+	return colblk.KeyComparison{
+		PrefixLen:         int32(prefixLen),
+		CommonPrefixLen:   int32(commonPrefixLen),
+		UserKeyComparison: int32(cmp.Compare(key[commonPrefixLen], lastRoachKey[commonPrefixLen])),
 	}
-	return cmpv
 }
 
 func (kw *cockroachKeyWriter) WriteKey(
@@ -165,9 +188,10 @@ func (kw *cockroachKeyWriter) WriteKey(
 	// TODO(jackson): Avoid copying the previous suffix.
 	kw.prevSuffix = append(kw.prevSuffix[:0], key[keyPrefixLen:]...)
 
-	// When the roach key is the same, keyPrefixLenSharedWithPrev includes the
-	// separator byte.
-	kw.roachKeys.Put(key[:keyPrefixLen-1], min(int(keyPrefixLenSharedWithPrev), int(keyPrefixLen)-1))
+	// When the roach key is the same or contain the previous key as a prefix,
+	// keyPrefixLenSharedWithPrev includes the previous key's separator byte.
+	kw.roachKeys.Put(key[:keyPrefixLen-1], min(int(keyPrefixLenSharedWithPrev), int(kw.prevRoachKeyLen)))
+	kw.prevRoachKeyLen = keyPrefixLen - 1
 
 	// NB: The w.logicalTimes builder was initialized with InitWithDefault, so
 	// if we don't set a value, the column value is implicitly zero. We only

--- a/pkg/storage/testdata/key_schema_key_writer
+++ b/pkg/storage/testdata/key_schema_key_writer
@@ -105,3 +105,35 @@ finish
 | hex:017a6b12706f690001 |          0 |       0 | 02073a83c45688420eaf97824255790f1e |
 | /MVCC/poi              | 1000000000 |       3 |                                    |
 +------------------------+------------+---------+------------------------------------+
+
+# Regression test for #134053.
+#
+# Write consecutive keys where the engine key prefix of the 1st key (including
+# the sentinel byte) is a byte prefix of the 2nd key, but the two keys have
+# different MVCC prefixes. Previously, ComparePrev would return the wrong value
+# for CommonPrefixLen, omitting the sentinel byte of the previous prefix.
+
+init
+----
+
+write
+hex:fa00180512db93969eee09
+hex:fa0000180512db93c71f7409
+----
+Parse("hex:fa00180512db93969eee09") = hex:fa00180512db93969eee09
+00: ComparePrev("hex:fa00180512db93969eee09"): PrefixLen=2; CommonPrefixLen=0; UserKeyComparison=1
+00: WriteKey(0, "hex:fa00180512db93969eee09", PrefixLen=2, CommonPrefixLen=0)
+00: MaterializeKey(_, 0) = hex:fa00180512db93969eee09
+Parse("hex:fa0000180512db93c71f7409") = hex:fa0000180512db93c71f7409
+01: ComparePrev("hex:fa0000180512db93c71f7409"): PrefixLen=3; CommonPrefixLen=2; UserKeyComparison=1
+01: WriteKey(1, "hex:fa0000180512db93c71f7409", PrefixLen=3, CommonPrefixLen=2)
+01: MaterializeKey(_, 1) = hex:fa0000180512db93c71f7409
+
+finish
+----
++----------+---------------------+---------+---------+
+|   KEY    |        WALL         | LOGICAL | UNTYPED |
++----------+---------------------+---------+---------+
+| hex:fa   | 1730810366077083374 |       0 |         |
+| hex:fa00 | 1730810366080262004 |       0 |         |
++----------+---------------------+---------+---------+

